### PR TITLE
replication: add GoCDK PubSub support

### PR DESCRIPTION
**Pull request by jba**

### CODE REVIEW

**Comment by chrislusf** [comment]
btw, the build failed currently. Seems transient, but similar problem as last time. I will try to restart the build.

```
$ go get ./weed/...
package gocloud.dev/pubsub: unrecognized import path "gocloud.dev/pubsub" (https fetch: Get http://gocloud.dev/pubsub/?go-get=1: redirected from secure URL https://gocloud.dev/pubsub?go-get=1 to insecure URL http://gocloud.dev/pubsub/?go-get=1)
package gocloud.dev/pubsub/awssnssqs: unrecognized import path "gocloud.dev/pubsub/awssnssqs" (https fetch: Get http://gocloud.dev/pubsub/awssnssqs/?go-get=1: redirected from secure URL https://gocloud.dev/pubsub/awssnssqs?go-get=1 to insecure URL http://gocloud.dev/pubsub/awssnssqs/?go-get=1)
package gocloud.dev/pubsub/azuresb: unrecognized import path "gocloud.dev/pubsub/azuresb" (https fetch: Get http://gocloud.dev/pubsub/azuresb/?go-get=1: redirected from secure URL https://gocloud.dev/pubsub/azuresb?go-get=1 to insecure URL http://gocloud.dev/pubsub/azuresb/?go-get=1)
package gocloud.dev/pubsub/gcppubsub: unrecognized import path "gocloud.dev/pubsub/gcppubsub" (https fetch: Get http://gocloud.dev/pubsub/gcppubsub/?go-get=1: redirected from secure URL https://gocloud.dev/pubsub/gcppubsub?go-get=1 to insecure URL http://gocloud.dev/pubsub/gcppubsub/?go-get=1)
package gocloud.dev/pubsub/natspubsub: unrecognized import path "gocloud.dev/pubsub/natspubsub" (https fetch: Get http://gocloud.dev/pubsub/natspubsub/?go-get=1: redirected from secure URL https://gocloud.dev/pubsub/natspubsub?go-get=1 to insecure URL http://gocloud.dev/pubsub/natspubsub/?go-get=1)
package gocloud.dev/pubsub/rabbitpubsub: unrecognized import path "gocloud.dev/pubsub/rabbitpubsub" (https fetch: Get http://gocloud.dev/pubsub/rabbitpubsub/?go-get=1: redirected from secure URL https://gocloud.dev/pubsub/rabbitpubsub?go-get=1 to insecure URL http://gocloud.dev/pubsub/rabbitpubsub/?go-get=1)
The command "go get ./weed/..." failed and exited with 1 during .
```

File: weed/replication/sub/notification_gocdk_pub_sub.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+type GoCDKPubSubInput struct {
+	sub *pubsub.Subscription
+}
+
+func (k *GoCDKPubSubInput) GetName() string {
+	return "gocdk_pub_sub"
+}
+
+func (k *GoCDKPubSubInput) Initialize(config util.Configuration) error {
+	subURL := config.GetString("sub_url")
```
Comment (chrislusf) [inline weed/replication/sub/notification_gocdk_pub_sub.go:31]: should this be "topic_url"?

**Comment by chrislusf** [comment]
the build still fails, only for go tip, not for go 1.12.x, 1.11.x, or 1.10.x

**Comment by chrislusf** [comment]
Need help on the redirect error. All builds are failing for go tips.

**Comment by jba** [comment]
I tried to reproduce. I couldn't.

File: weed/replication/sub/notification_gocdk_pub_sub.go (hunk 1 above)
Comment (jba) [reply weed/replication/sub/notification_gocdk_pub_sub.go:31]: No, for receiving we want a subscription.

There was a typo in the log message. Fixed.

**Comment by jba** [comment]
Will keep trying to repro.

**Comment by jba** [comment]
Reproduced. This is a known issue: golang/go#31241. 

Not sure why you want to test at tip, since it isn't stable. But we can just leave this PR here until they fix it, which should be soon.

File: weed/replication/sub/notification_gocdk_pub_sub.go (hunk 1 above)
Comment (chrislusf) [reply weed/replication/sub/notification_gocdk_pub_sub.go:31]: seems this is different from existing "notification.gocdk_pub_sub.topic_url". Is a "notification.gocdk_pub_sub.sub_url" needed in the scaffold?

https://github.com/chrislusf/seaweedfs/blob/master/weed/command/scaffold.go#L192
https://github.com/chrislusf/seaweedfs/blob/master/weed/notification/gocdk_pub_sub/gocdk_pub_sub.go#L47

Comment (jba) [reply weed/replication/sub/notification_gocdk_pub_sub.go:31]: Thanks for noticing. I added an entry.

Comment (chrislusf) [reply weed/replication/sub/notification_gocdk_pub_sub.go:31]: Actually, this also need to change? I suppose the topic_url and sub_url are referring to the same thing.

https://github.com/chrislusf/seaweedfs/blob/master/weed/notification/gocdk_pub_sub/gocdk_pub_sub.go#L47

Comment (jba) [reply weed/replication/sub/notification_gocdk_pub_sub.go:31]: They aren't referring to the same thing. In many Pub/Sub systems (Google Pub/Sub, Azure Service Bus, etc.) there are two separate concepts: topics and subscriptions. You publish to a topic. You receive from a subscription. They have different names, and usually must be created and deleted separately. When you create a subscription, you name the topic that it receives from, but other than that they are independent resources.

**Comment by chrislusf** [comment]
Thanks! Merged!

Another related question: when using https://github.com/jondot/goweight to analyze the binary, the top one is `github.com/ugorji/go/codec`, coming from "gocloud.dev/pubsub/natspubsub". 11MB is a large portion of 59MB generated binary.

```
weed$ goweight
   11 MB github.com/ugorji/go/codec
  6.1 MB github.com/aws/aws-sdk-go/service/s3
  5.2 MB github.com/Azure/azure-storage-blob-go/azblob
  4.3 MB github.com/go-redis/redis
  4.1 MB github.com/DataDog/zstd
  3.8 MB net/http
  3.5 MB github.com/Shopify/sarama
  3.3 MB runtime
  3.0 MB github.com/gocql/gocql
  1.9 MB golang.org/x/net/http2
  1.8 MB github.com/golang/protobuf/proto
  1.8 MB net
  1.8 MB crypto/tls
```

**Comment by jba** [comment]
Removed igorji/go/codec. See google/go-cloud#1776.

**Comment by chrislusf** [comment]
Thanks! Works well! The total size 59732788 bytes=> 55997588 bytes, about 3.56MB, or 6.25% reduction.

The new goweight output:
```
weed$ goweight
  6.1 MB github.com/aws/aws-sdk-go/service/s3
  5.5 MB github.com/DataDog/zstd
  5.4 MB github.com/Azure/azure-storage-blob-go/azblob
  4.3 MB github.com/go-redis/redis
  3.8 MB net/http
  3.6 MB github.com/Shopify/sarama
  3.3 MB runtime
  3.0 MB github.com/gocql/gocql
  1.9 MB golang.org/x/net/http2
  1.8 MB github.com/golang/protobuf/proto
  1.8 MB net
  1.8 MB crypto/tls
```

**Pull request merged**